packages/common/container-definitions/src/loader.ts

```diff
@@ -102,23 +102,134 @@ export interface ICodeAllowList {
 export interface IContainerEvents extends IEvent {
+    /**
+     * Emitted when the readonly state of the container changes.
+     *
+     * @remarks Listener parameters:
+     *
+     * - `readonly`: Whether or not the container is now in a readonly state.
+     *
+     * @see {@link IContainer.readOnlyInfo}
+     */
     (event: "readonly", listener: (readonly: boolean) => void): void;
+
+    /**
+     * Emitted when the {@link IContainer} completes connecting to the Fluid service.
+     *
+     * @remarks Reflects connection state changes against the (delta) service acknowledging ops/edits.
+     *
+     * @see
+     *
+     * - {@link IContainer.connectionState}
+     *
+     * - {@link IContainer.connect}
+     */
     (event: "connected", listener: (clientId: string) => void);
+
+    /**
+     * TODO
+     *
+     * @remarks Listener parameters:
+     *
+     * - `codeDetails`: TODO
+     *
+     * - `proposal`: TODO
+     *
+     * @see {@link IContainer.proposeCodeDetails}
+     */
     (event: "codeDetailsProposed", listener: (codeDetails: IFluidCodeDetails, proposal: ISequencedProposal) => void);
+
+    /**
+     * TODO
+     *
+     * @remarks Listener parameters:
+     *
+     * - `codeDetails`: TODO
+     */
     (event: "contextChanged", listener: (codeDetails: IFluidCodeDetails) => void);
-    (event: "disconnected" | "attached", listener: () => void);
+
+    /**
+     * Emitted when the {@link IContainer} becomes disconnected from the Fluid service.
+     *
+     * @remarks Reflects connection state changes against the (delta) service acknowledging ops/edits.
+     *
+     * @see
+     *
+     * - {@link IContainer.connectionState}
+     *
+     * - {@link IContainer.disconnect}
+     */
+    (event: "disconnected", listener: () => void);
+
+    /**
+     * TODO
+     *
+     * @see {@link IContainer.attachState}
+     */
+    (event: "attached", listener: () => void);
+
+    /**
+     * Emitted when the {@link IContainer} is closed, which permanently disables it.
+     *
+     * @remarks Listener parameters:
+     *
+     * - `error`: If the container was closed due to error, this will contain details about the error that caused it.
+     */
     (event: "closed", listener: (error?: ICriticalContainerError) => void);
+
+    /**
+     * TODO
+     *
+     * @remarks Listener parameters:
+     *
+     * - `error`: TODO
+     */
     (event: "warning", listener: (error: ContainerWarning) => void);
+
+    /**
+     * TODO
+     *
+     * @remarks Listener parameters:
+     *
+     * - `message`: TODO
+     */
     (event: "op", listener: (message: ISequencedDocumentMessage) => void);
-    (event: "dirty" | "saved", listener: (dirty: boolean) => void);
+
+    /**
+     * Emitted when the first local change has been made, following a "saved" event.
+     *
+     * @remarks
+     *
+     * The "saved" event will be emitted once all local changes have been acknowledged by the service.
```

Comment:
Looks like you swapped "saved" with "dirty" here and below.

Reply:
This was intentional. It's sort of an aside, noting how the *other* event relates to this one. Unfortunately, TSDoc syntax does not currently support links to call-signatures, or I would actually link.

That said, this intention was obviously not clear, since it created confusion. Any suggestions on what would make this more clear?

Reply:
Oh I see.  This is very event-centric documentation - perhaps it would work better to rely on defining the states on isDirty and then the event documentation just refers to that (fires when transitioning to dirty, fires when transitioning to saved)

Reply:
I like that. Will update.